tests/components/gree/test_climate.py

```diff
@@ -0,0 +1,781 @@
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    with pytest.raises(ValueError):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_SET_SWING_MODE,
+            {ATTR_ENTITY_ID: ENTITY_ID, ATTR_SWING_MODE: "invalid"},
+            blocking=True,
+        )
+
+    state = hass.states.get(ENTITY_ID)
+    assert state is not None
+    assert state.attributes.get(ATTR_SWING_MODE) != "invalid"
+
+
+@pytest.mark.parametrize(
+    "swing_mode", (SWING_OFF, SWING_BOTH, SWING_VERTICAL, SWING_HORIZONTAL)
+)
+async def test_send_swing_mode_device_timeout(
+    hass, discovery, device, mock_now, swing_mode
+):
+    """Test for sending swing mode command to the device with a device timeout."""
+    device().push_state_update.side_effect = DeviceTimeoutError
+
+    await async_setup_gree(hass)
+
+    next_update = mock_now + timedelta(minutes=5)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    assert await hass.services.async_call(
+        DOMAIN,
+        SERVICE_SET_SWING_MODE,
+        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_SWING_MODE: swing_mode},
+        blocking=True,
+    )
+
+    state = hass.states.get(ENTITY_ID)
+    assert state is not None
+    assert state.attributes.get(ATTR_SWING_MODE) == swing_mode
+
+
+@pytest.mark.parametrize(
+    "swing_mode", (SWING_OFF, SWING_BOTH, SWING_VERTICAL, SWING_HORIZONTAL)
+)
+async def test_update_swing_mode(hass, discovery, device, mock_now, swing_mode):
+    """Test for updating swing mode from the device."""
+    device().horizontal_swing = (
+        HorizontalSwing.FullSwing
+        if swing_mode in (SWING_BOTH, SWING_HORIZONTAL)
+        else HorizontalSwing.Default
+    )
+    device().vertical_swing = (
+        VerticalSwing.FullSwing
+        if swing_mode in (SWING_BOTH, SWING_VERTICAL)
+        else VerticalSwing.Default
+    )
+
+    await async_setup_gree(hass)
+
+    next_update = mock_now + timedelta(minutes=5)
+    with patch("homeassistant.util.dt.utcnow", return_value=next_update):
+        async_fire_time_changed(hass, next_update)
+    await hass.async_block_till_done()
+
+    state = hass.states.get(ENTITY_ID)
+    assert state is not None
+    assert state.attributes.get(ATTR_SWING_MODE) == swing_mode
+
+
+async def test_name(hass, discovery, device):
+    """Test for name property."""
+    await async_setup_gree(hass)
+    state = hass.states.get(ENTITY_ID)
+    assert state.attributes[ATTR_FRIENDLY_NAME] == "fake-device-1"
+
+
+async def test_supported_features_with_turnon(hass, discovery, device):
+    """Test for supported_features property."""
+    await async_setup_gree(hass)
+    state = hass.states.get(ENTITY_ID)
+    assert state.attributes[ATTR_SUPPORTED_FEATURES] == SUPPORTED_FEATURES
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6333224)